### PR TITLE
Add a container type to easily hide online content when not online

### PR DESCRIPTION
**Pull request by Game4all**
# Summary

- This PR implements the `OnlineViewContainer`, a container for displaying online-only views where the user login is required.
- It supports hiding its content and showing a login placeholder (implemented in #7438) if the local user is not logged in
- It also displays a loading animation when the user is connecting to game servers.
- This component is planned to be reused to refine UX in online-only overlays (Social, Direct, Chat) instead of displaying an empty overlay

### CODE REVIEW

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 1]
```diff
@@ -151,7 +151,7 @@ protected PlaceholderState PlaceholderState
                         break;
 
                     case PlaceholderState.NotLoggedIn:
-                        replacePlaceholder(new LoginPlaceholder());
+                        replacePlaceholder(new LoginPlaceholder(@"Please sign in to view online leaderboards!"));
```
Comment (iiSaLMaN) [inline osu.Game/Online/Leaderboards/Leaderboard.cs]: This case should probably be removed and inherit `OnlineViewContainer`.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                placeholderContainer = new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Alpha = 0,
+                    Child = placeholder = new LoginPlaceholder(placeholder_message)
+                },
+            };
+        }
+
+        public void APIStateChanged(IAPIProvider api, APIState state)
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: Should be `virtual` to allow components that inherit this and also want `IOnlineComponent` features to override this.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            }
+            else
+            {
+                placeholderContainer.FadeOut(transform_time / 2, Easing.OutQuint);
+                content.FadeIn(transform_time, Easing.OutQuint);
+            }
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(IAPIProvider api)
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: I think this should be a protected property
```csharp
[Resolved]
protected IAPIProvider API { get; set; }
```

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        private OnlineViewContainer onlineView;
+        private Box box;
+
+        public TestSceneOnlineViewContainer()
+        {
+            Child = new Container
+            {
+                RelativeSizeAxes = Axes.Both,
+                Child = onlineView = new OnlineViewContainer(@"Please login to view dummy test content")
```
Comment (mcendu) [inline osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: Why not “sign in”?

File: osu.Game/Online/OnlineViewContainer.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+using osu.Game.Online.API;
+using osu.Game.Online.Placeholders;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A <see cref="Container"/> for dislaying online content who require a local user to be logged in.
+    /// Shows its children only when the local user is logged in and supports displaying a placeholder if not.
+    /// </summary>
+    public class OnlineViewContainer : Container, IOnlineComponent
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: Should obviously be `abstract`.

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs (hunk 4 above)
Comment (iiSaLMaN) [reply osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: One thing I was aware from this is the placeholder message becoming inconsistent across the overlays by simply one of them having `log in` instead of `sign in`, I think the string should be concatenated somewhere inside the `LoginPlaceholder` in a way that the developer only need to pass reason to log in: 
```
OnlineViewContainer(@"dummy test content") -> $@"Please sign in to view {actionMessage}" -> "Please sign in to view dummy test content"
```

Comment (Game4all) [reply osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: This was in the initial implementation of the login placeholder but has been reverted to passing the whole string because string concatenation was deemed bad for localization.

Comment (iiSaLMaN) [reply osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: I mean, it will be concatenated before locallizing, at the constructor of `LoginPlaceholder`.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                placeholderContainer.FadeOut(transform_time / 2, Easing.OutQuint);
+                content.FadeIn(transform_time, Easing.OutQuint);
+            }
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(IAPIProvider api)
+        {
+            api.Register(this);
```
[the remark this answers is not in the document]
Comment (Game4all) [reply osu.Game/Online/OnlineViewContainer.cs]: I don't see in what this is a memory leak, this is being done in several places without leaking any memory (see `FullscreenOverlay`)

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs (hunk 4 above)
Comment (Game4all) [reply osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: Sounds fine in that case.

File: osu.Game/Online/OnlineViewContainer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (iiSaLMaN) [reply osu.Game/Online/OnlineViewContainer.cs]: Well, It doesn't leak there because it's an overlay, which I believe never removed and re-created, but this is a container, although as a rule of thumb, you should unregister the component from any place it's registered at disposal.

[the remark this answers is not in the document]
Comment (Game4all) [reply osu.Game/Online/OnlineViewContainer.cs]: You're completely right , my bad.

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 1 above)
Comment (Game4all) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: Should be kept for a follow-up PR covering usages  imo.

File: osu.Game/Online/OnlineViewContainer.cs (hunk 5 above)
Comment (Game4all) [reply osu.Game/Online/OnlineViewContainer.cs]: Not sure about that one, there are place where we may want to use the container as is, without need to subclass

File: osu.Game/Online/OnlineViewContainer.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    /// </summary>
+    public class OnlineViewContainer : Container, IOnlineComponent
+    {
+        private readonly Container content;
+        private readonly Container placeholderContainer;
+        private readonly Placeholder placeholder;
+
+        private const int transform_time = 300;
+
+        protected override Container<Drawable> Content => content;
```
Comment (bdach) [inline osu.Game/Online/OnlineViewContainer.cs]: This and the underlying field is a bit confusing. An improperly-written derived class could override this get-only protected property to return something different, but have to call the base ctor, in which case `content` is initialised in the private field in the base, but not accessible to the derived, therefore creating one dangling container reference per instance. You probably want a protected auto property here.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                },
+            };
+        }
+
+        public virtual void APIStateChanged(IAPIProvider api, APIState state)
+        {
+            switch (state)
+            {
+                case APIState.Offline:
+                case APIState.Connecting:
```
Comment (bdach) [inline osu.Game/Online/OnlineViewContainer.cs]: Are we sure we want the offline placeholder visible during connecting, instead of a loading animation of sorts?

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: In this case, why is this change even here? It seems orthogonal and unrelated.

File: osu.Game/Online/Placeholders/LoginPlaceholder.cs [hunk 9]
```diff
@@ -14,15 +14,15 @@ public sealed class LoginPlaceholder : Placeholder
...
+        public LoginPlaceholder(string actionMessage)
         {
             AddIcon(FontAwesome.Solid.UserLock, cp =>
             {
                 cp.Font = cp.Font.With(size: TEXT_SIZE);
                 cp.Padding = new MarginPadding { Right = 10 };
             });
 
-            AddText(@"Please sign in to view online leaderboards!");
+            AddText(actionMessage);
```
Comment (iiSaLMaN) [inline osu.Game/Online/Placeholders/LoginPlaceholder.cs]: I meant the concatenation would reside here:
`AddText($"Please sign in to {actionMessage}!");`

File: osu.Game/Online/OnlineViewContainer.cs [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            base.LoadComplete();
+        }
+
+        protected override void Dispose(bool isDisposing)
+        {
+            API?.Unregister(this);
+            base.Dispose(isDisposing);
+        }
+
+        protected enum PlaceholderStatus
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: Also as a rule in `enum` names, they should not be plural (`Status` is plural), you probably want `State`.

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 11]
```diff
@@ -166,6 +161,7 @@ protected PlaceholderState PlaceholderState
         }
 
         protected Leaderboard()
+            : base("view online leaderboards!")
         {
             InternalChildren = new Drawable[]
```
Comment (iiSaLMaN) [inline osu.Game/Online/Leaderboards/Leaderboard.cs]: Have you tested this correctly, it seems that all internal children of the `OnlineViewContainer` got ripped out.

File: osu.Game/Online/Leaderboards/PlaceholderState.cs [hunk 12]
```diff
@@ -11,7 +11,6 @@ public enum PlaceholderState
         Unavailable,
         NoneSelected,
         NoScores,
-        NotLoggedIn,
```
Comment (iiSaLMaN) [inline osu.Game/Online/Leaderboards/PlaceholderState.cs]: `PlaceholderState` used for `Leaderboard<>` should probably be renamed to `LeaderboardPlaceholderState` for now.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        {
+            Offline,
+            Online,
+            Connecting,
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: The namings in this enum seem out-of-scope, what you probably want here is:
 - `Visible`
 - `Hidden`
 - `Loading`

with xmldocs shortly explaining what these status mean to the container.

Comment (iiSaLMaN) [reply osu.Game/Online/OnlineViewContainer.cs]: Reason for this is to expand the container into supporting more than just 1 (not logged in) placeholder.

**Review by frenzibyte — CHANGES_REQUESTED** [review]
Seems like `Leaderboard<>` needs much work to inherit `OnlineViewContainer` correctly and use it nicely, I advocate you revert the inheritance and remove any change to it and leave it as a follow-up PR.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                    Schedule(() => UpdatePlaceholderVisibility(PlaceholderStatus.Offline));
+                    break;
+
+                case APIState.Online:
+                    Schedule(() => UpdatePlaceholderVisibility(PlaceholderStatus.Online));
+                    break;
+            }
+        }
+
+        protected void UpdatePlaceholderVisibility(PlaceholderStatus status)
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: Why not use `APIState` directly?

File: osu.Game/Online/Placeholders/LoginPlaceholder.cs (hunk 9 above)
Comment (peppy) [reply osu.Game/Online/Placeholders/LoginPlaceholder.cs]: This will not localise well. Fine as-is.

File: osu.Game/Online/OnlineViewContainer.cs (hunk 10 above)
Comment (peppy) [reply osu.Game/Online/OnlineViewContainer.cs]: why not just use apistate

File: osu.Game/Online/Leaderboards/Leaderboard.cs [hunk 15]
```diff
@@ -166,6 +161,7 @@ protected PlaceholderState PlaceholderState
         }
 
         protected Leaderboard()
+            : base("view online leaderboards!")
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: This shouldn't be done in the first place as it can't be localised. The full message should be constructed for each usage.

[the remark this answers is not in the document]
Comment (iiSaLMaN) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: Correct me if i'm wrong, but shouldn't the localising happen inside `LocalisedString`? 
Which means `SpriteText.Text` is the property that should receive the full message. (which should happen in `LoginPlaceholder`)

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: It's out of scope of this PR, but have a think how you would generate such localisation strings and how (with the logic you suggest) that can't be done statically.

**Comment by Game4all** [comment]
Will revert any changes made to leaderboard class and apply them in a separate PR as it requires much more work into it as stated by @iiSaLMaN.

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (Game4all) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: So, also reverting string concatenation in constructor

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                        }
+                    }
+                }
+            };
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
```
Comment (iiSaLMaN) [inline osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: why not a test method
```csharp
[Test]
public void TestVisibilityBehaviour()
{
    ...
}
```

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            AddStep("set status to offline", () => ((DummyAPIAccess)API).State = APIState.Offline);
+
+            AddAssert("children are hidden", () => !onlineView.Children.First().Parent.IsPresent);
```
Comment (iiSaLMaN) [inline osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: I advocate you expose `View` instead of playing with the sub-tree to retrieve it.

also, checking is present is not the best way to determine if a drawable is hidden, it may have a `override IsPresent => true` or custom values because it's a manager or something, check `Alpha` directly instead.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+        /// <summary>
+        /// Updates the visibility of the placeholder for this <see cref="OnlineViewContainer"/> accordingly to the passed <see cref="APIState"/>.
+        /// <list type="bullet">
+        /// <item> Passing <see cref="APIState.Online"/> will cause the online-only view to display.</item>
+        /// <item> Passing <see cref="APIState.Offline"/> will cause the placeholder to display.</item>
+        /// <item> Passing <see cref="APIState.Connecting"/> or <see cref="APIState.Failing"/> will cause the loading animation to display.</item>
+        /// </list>
+        /// </summary>
+        /// <param name="status">The new API state</param>
+        protected void UpdatePlaceholderVisibility(APIState status)
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: Is there a reason this is protected? I'm not sure if it's fine for a deriver to be able to manipulate the placeholder visibility.

**Review by frenzibyte — CHANGES_REQUESTED** [review]
Seems OK mostly except for a thing, the rest is all on the test. (the place I missed in the last review)

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs (hunk 17 above)
Comment (Game4all) [reply osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: Actually, since View is a normal container, checking if drawable is visible through IsPresent should be safe

File: osu.Game/Online/OnlineViewContainer.cs [hunk 19]
```diff
@@ -18,7 +18,7 @@ public class OnlineViewContainer : Container, IOnlineComponent
     {
         private readonly Container placeholderContainer;
         private readonly Placeholder placeholder;
-        private readonly LoadingAnimation loading;
+        internal readonly LoadingAnimation LoadingAnimation;
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: Doesn't need to be that exposed, what you're looking for is making this protected and expose it on test sub-class with `public new LoadingAnimation => base.LoadingAnimation`.

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs (hunk 17 above)
Comment (iiSaLMaN) [reply osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: Well, using `IsPresent` is technically not safe because you're not checking alpha, but you're checking a property that has alpha in it's calculation, I find it more arbitrary.

Comment (Game4all) [reply osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: Searching for IsPresent in Tests yields results where that property is being used to check if a drawable is displayed on screen. So it should be fine to use it to check for drawable visibility
Capture

File: osu.Game/Online/OnlineViewContainer.cs [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+    /// </summary>
+    public class OnlineViewContainer : Container, IOnlineComponent
+    {
+        private readonly Container placeholderContainer;
+        private readonly Placeholder placeholder;
+        protected readonly LoadingAnimation LoadingAnimation;
+
+        private const int transform_time = 300;
+
+        protected Container View { get; }
```
Comment (peppy) [inline osu.Game/Online/OnlineViewContainer.cs]: combine with content.

`protected override Container<Drawable> Content { get; }`

File: osu.Game/Online/OnlineViewContainer.cs [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+                    Child = placeholder = new LoginPlaceholder(placeholderMessage)
+                },
+                LoadingAnimation = new LoadingAnimation
+                {
+                    Alpha = 0,
+                }
+            };
+        }
+
+        public virtual void APIStateChanged(IAPIProvider api, APIState state) => updatePlaceholderVisibility(state);
```
Comment (peppy) [inline osu.Game/Online/OnlineViewContainer.cs]: combine with private method below. it's only called from here.

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 1 above)
Comment (peppy) [inline osu.Game/Online/Leaderboards/Leaderboard.cs]: seems unrelated to this implementation.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Online/OnlineViewContainer.cs [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+using osu.Game.Online.API;
+using osu.Game.Online.Placeholders;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A <see cref="Container"/> for dislaying online content who require a local user to be logged in.
+    /// Shows its children only when the local user is logged in and supports displaying a placeholder if not.
+    /// </summary>
+    public class OnlineViewContainer : Container, IOnlineComponent
```
Comment (peppy) [inline osu.Game/Online/OnlineViewContainer.cs]: should be abstract.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Online/Leaderboards/Leaderboard.cs (hunk 1 above)
Comment (iiSaLMaN) [reply osu.Game/Online/Leaderboards/Leaderboard.cs]: it is actually, the `LoginPlaceholder` constructor now has a message parameter.

File: osu.Game/Online/OnlineViewContainer.cs (hunk 20 above)
Comment (Game4all) [reply osu.Game/Online/OnlineViewContainer.cs]: Reverting suggested behaviour as it causes the container to not exhibit desired behaviour (Content doesn't hide) when `Content` property is overriden in sub-class

File: osu.Game/Online/OnlineViewContainer.cs [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+
+        protected OnlineViewContainer(string placeholderMessage)
+        {
+            InternalChildren = new Drawable[]
+            {
+                content = new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                },
+                placeholderContainer = new Container
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: any reason why this placeholder is added in a container? (tested removing it and animating everything directly in `placeholder`, no issues.)

File: osu.Game/Online/OnlineViewContainer.cs [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+                }
+            };
+        }
+
+        public virtual void APIStateChanged(IAPIProvider api, APIState state)
+        {
+            switch (state)
+            {
+                case APIState.Offline:
+                    content.FadeOut(150, Easing.OutQuint);
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: This is limiting this container into just fading out on placeholder appearance, would be much better to make this customizable (by inheriting `VisibilityContainer` and setting visibility here) and add default behaviour here (fading in & out) with custom usages included in tests. (use grid container to show them both)

**Review by frenzibyte — CHANGES_REQUESTED** [review]
Tested, seems pretty good otherwise.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+                }
+            };
+        }
+
+        public virtual void APIStateChanged(IAPIProvider api, APIState state)
+        {
+            switch (state)
+            {
+                case APIState.Offline:
+                    PopOut();
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: That's a very dangerous way to do it (visibility state is not updated), you should be calling `Hide()` and `Show()` instead.

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+                    {
+                        Text = "dummy online content",
+                        RelativeSizeAxes = Axes.Both,
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                    }
+                };
+            }
+
+            protected override void PopOut() => Content.FadeOut(150, Easing.OutQuint);
```
Comment (iiSaLMaN) [inline osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: another way to show that it's customizable is by creating a `BufferedContainer` and blur it here.

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs (hunk 26 above)
Comment (Game4all) [reply osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: Simply indicating in class XMLDoc that PopIn() and PopOut() can be overridden to customize fade behaviour should be sufficient imo

File: osu.Game/Online/OnlineViewContainer.cs [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Online.API;
+using osu.Game.Online.Placeholders;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A <see cref="Container"/> for dislaying online content who require a local user to be logged in.
+    /// Shows its children only when the local user is logged in and supports displaying a placeholder if not.
+    /// <see cref="PopIn"/> and <see cref="PopOut"/> methods can be overriden to customize fade in/out behaviour
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]:
```suggestion
    /// <see cref="PopIn"/> and <see cref="PopOut"/> methods can be overriden to customize visibility animation.
```

Comment (peppy) [reply osu.Game/Online/OnlineViewContainer.cs]: why is this necessary in the first place? that should be documented on `VisibilityContainer`, not this one.

File: osu.Game/Online/OnlineViewContainer.cs (hunk 20 above)
Comment (peppy) [reply osu.Game/Online/OnlineViewContainer.cs]: i don't understand.

Comment (Game4all) [reply osu.Game/Online/OnlineViewContainer.cs]: Actually, nevermind it can be merged back. 
Care just needs to be taken to call `base.Content` when customizing the visibility animations of a drawable inheriting `OnlineViewContainer` and also overriding `Content` property. Else it would
cause the online view to still be displayed even if it should be hidden (as seen with a refactor of `Leaderboard<>` class inheriting `OnlineViewContainer`)

image

Comment (peppy) [reply osu.Game/Online/OnlineViewContainer.cs]: The component should be made in a way that this is not possible. Also, the change proposed here doesn't affect this – it could still happen with your current code.

**Comment by Game4all** [comment]
I did a bit of cleanup in this PR commit history by removing commits of my past attempts at making Leaderboard derive from OnlineViewContainer. Should be ready for a re-review.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+                    FadeContentOut(TransformationTarget);
+                    LoadingAnimation.Show();
+                    placeholder.FadeOut(TRANSFORM_TIME / 2, Easing.OutQuint);
+                    break;
+            }
+        }
+
+        protected abstract void FadeContentOut(Drawable content);
+
+        protected abstract void FadeContentIn(Drawable content);
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: The naming seems to be unnecessarily specific (content may not be faded, but blurred or darkened), how about `PopContentIn` and `PopContentOut`, or `AnimateHide` and `AnimateShow`?

File: osu.Game/Online/OnlineViewContainer.cs [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+    /// Shows its children only when the local user is logged in and supports displaying a placeholder if not.
+    /// </summary>
+    public abstract class OnlineViewContainer : Container, IOnlineComponent
+    {
+        private readonly Placeholder placeholder;
+        protected readonly LoadingAnimation LoadingAnimation;
+
+        protected const double TRANSFORM_TIME = 300.0;
+
+        internal readonly Container TransformationTarget;
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: why `internal` (if for testing purposes, never expose a property in normal classes for testing purposes).

Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: would be better to rename this to `ViewTarget`.

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs [hunk 30]
```diff
@@ -0,0 +1,93 @@
...
+            AddStep("set status to connecting", () => ((DummyAPIAccess)API).State = APIState.Connecting);
+
+            AddAssert("children are hidden", () => !onlineView.TransformationTarget.IsPresent);
+            AddAssert("loading animation is visible", () => onlineView.LoadingAnimation.IsPresent);
+
+            AddStep("set status to failing", () => ((DummyAPIAccess)API).State = APIState.Failing);
+
+            AddAssert("children are hidden", () => !onlineView.TransformationTarget.IsPresent);
+            AddAssert("loading animation is visible", () => onlineView.LoadingAnimation.IsPresent);
+        }
```
Comment (bdach) [inline osu.Game.Tests/Visual/Online/TestSceneOnlineViewContainer.cs]: I'd split this testcase in two (`TestConnectingStateVisibility` and `TestFailingStateVisibility`). You could even make all four test cases into a parameterised test for brevity but I guess that's just extra points.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Online.API;
+using osu.Game.Online.Placeholders;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A <see cref="Container"/> for dislaying online content which require a local user to be logged in.
```
Comment (bdach) [inline osu.Game/Online/OnlineViewContainer.cs]:
```suggestion
    /// A <see cref="Container"/> for displaying online content which require a local user to be logged in.
```

File: osu.Game/Online/OnlineViewContainer.cs (hunk 29 above)
Comment (bdach) [reply osu.Game/Online/OnlineViewContainer.cs]: `protected` with exposition in `TestSceneOnlineViewContainer` could work, but then it's kind of not great how you can do things with `TransformationTarget` on your own and *also* define `FadeContent{In,Out}`. Maybe should be `private` instead, and the tests should check the `Box` itself with `ChildrenOfType<>`. Not 100% sure either way.

Comment (iiSaLMaN) [reply osu.Game/Online/OnlineViewContainer.cs]: right, `TransformationTarget` should not be visible not even to the sub-class, checking *content* of the `TestOnlineViewContainer` is probably the best, I advocate for that.

File: osu.Game/Online/OnlineViewContainer.cs [hunk 32]
```diff
@@ -21,7 +21,7 @@ public abstract class OnlineViewContainer : Container, IOnlineComponent
 
         protected const double TRANSFORM_TIME = 300.0;
 
-        internal readonly Container TransformationTarget;
+        private readonly Container viewTarget;
         protected override Container<Drawable> Content { get; }
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: why not extract the backing-field of this property and using it as the "view target", adding a second layer seems redundant.
```csharp
private readonly Container content;
protected override Container<Drawable> Content => content;
```

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game/Online/OnlineViewContainer.cs [hunk 33]
```diff
@@ -0,0 +1,88 @@
...
+                    PopContentOut(viewContent);
+                    LoadingAnimation.Show();
+                    placeholder.FadeOut(TRANSFORM_TIME / 2, Easing.OutQuint);
+                    break;
+            }
+        }
+
+        protected abstract void PopContentOut(Drawable content);
+
+        protected abstract void PopContentIn(Drawable content);
```
Comment (iiSaLMaN) [inline osu.Game/Online/OnlineViewContainer.cs]: Why did this become abstract? Most of the subclasses may use basic fading animation, that's what `virtual` is mostly useful for.

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game/Online/OnlineViewContainer.cs (hunk 33 above)
Comment (bdach) [reply osu.Game/Online/OnlineViewContainer.cs]: it always was abstract as far as I can see, FWIW

File: osu.Game/Online/OnlineViewContainer.cs [hunk 34]
```diff
@@ -0,0 +1,88 @@
...
+                    PopContentOut(viewContent);
+                    LoadingAnimation.Show();
+                    placeholder.FadeOut(TRANSFORM_TIME / 2, Easing.OutQuint);
+                    break;
+            }
+        }
+
+        protected virtual void PopContentOut(Drawable content) => content.FadeOut(TRANSFORM_TIME / 2, Easing.OutQuint);
+
+        protected virtual void PopContentIn(Drawable content) => content.FadeIn(TRANSFORM_TIME, Easing.OutQuint);
```
Comment (bdach) [inline osu.Game/Online/OnlineViewContainer.cs]: Strange how this is the only place that uses `TRANSFORM_TIME` instead of this halved. Did you really mean this, or did you want

```suggestion
        protected virtual void PopContentIn(Drawable content) => content.FadeIn(TRANSFORM_TIME / 2, Easing.OutQuint);
```

here too?

**Review by bdach — COMMENTED** [review]
Save from one arbitrary time value I think this is ready to go. Let's get this in and move on.

File: osu.Game/Online/OnlineViewContainer.cs (hunk 33 above)
Comment (iiSaLMaN) [reply osu.Game/Online/OnlineViewContainer.cs]: it was overridden when it inherited `VisibilityContainer`.

File: osu.Game/Online/OnlineViewContainer.cs (hunk 34 above)
Comment (peppy) [reply osu.Game/Online/OnlineViewContainer.cs]: i believe this was intentional. generally feels good to fade out at double the speed of the related fade in

File: osu.Game/Online/OnlineViewContainer.cs [hunk 35]
```diff
@@ -0,0 +1,88 @@
...
+                case APIState.Failing:
+                case APIState.Connecting:
+                    PopContentOut(viewContent);
+                    LoadingAnimation.Show();
+                    placeholder.FadeOut(TRANSFORM_TIME / 2, Easing.OutQuint);
+                    break;
+            }
+        }
+
+        protected virtual void PopContentOut(Drawable content) => content.FadeOut(TRANSFORM_TIME / 2, Easing.OutQuint);
```
Comment (peppy) [inline osu.Game/Online/OnlineViewContainer.cs]: not sure these need to exist until they have a use case, but i guess no harm..

**Comment by peppy** [comment]
@Game4all i've made a few code quality and a few more important fixes (like potentially failing test steps). please check my commits for future contributions!

**Review by peppy — APPROVED** [review]
